docs/sources/architecture/ring.md

```diff
@@ -0,0 +1,78 @@
+
+Interestingly, this approach solves some of the problem we have, say same log stream goes to same server (because same set of labels + tenant gives same hash value), and while reading, we can hash it back to find the server where its value is stored.
+
+But this solution lacks some scaling properties. Say if we introduce new server or remove existing server (intentionally or unintentially) then every single log stream will map to different server now.
+
+There are several better approach to solve this problem. One of the approach we use in Loki is Consistent Hashing.
+
+## How it works
+
+Distributors use consistent hashing in conjunction with a configurable replication factor to determine which instances of the ingester service should receive a given stream.
+
+Each ingester holds a subset of tokens within the hash ring. This hash ring stored in Consul is used to achieve consistent hashing;
+
+Every ingester (also distributor) that is part of the ring has two things associated with it.
+- State
+- Set of tokens they own
+
+State can be one of the following
+- PENDING
+- JOINING
+- ACTIVE
+- LEAVING
+- UNHEALTHY
+
+The state ACITVE may receive both read and write requests. While state JOINING can receive only write requests, the state LEAVING may receive read requests.
+
+Each token is a random unsigned 32-bit number.
+
+When doing a hash lookup, distributors only use tokens for ingesters who are in the appropriate state for the request.
+
+To do the hash lookup, distributors find the smallest appropriate token whose value is larger than the hash of the stream. When the replication factor is larger than 1, the next subsequent tokens (clockwise in the ring) that belong to different ingesters will also be included in the result.
+
+The effect of this hash set up is that each token that an ingester owns is responsible for a range of hashes (hence load is distributed more evenly across the instances). If there are three tokens with values 0, 25, and 50, then a hash of 3 would be given to the ingester that owns the token 25; the ingester owning token 25 is responsible for the hash range of 1-25.
+
+Token is same as virtual node (or vnode) in consistent hashing.
+
+Number of tokens per ring is configurable via `--ingester.num-tokens`. Default is 128
+
+### Quorum consistency
+Since all ingesters share access to the same hash ring, write requests can be sent to any ingester.
+
+To ensure consistent query results, Loki uses Dynamo-style quorum consistency on reads and writes. This means for example that the distributor will wait for a positive response of at least one half plus one of the ingesters to send the sample to before responding to the client that initiated the send.
+
+### Ring Status page
+
+If Loki is run in microservice mode, both ingester and distributor exposes their ring status via the endpoint `api/v1/ruler/ring`
```

Comment:
This endpoint is only the ruler ring. The ingester ring is stored at `/ring` and that's the one I'd mention here. Followup: we should really expose all the rings on their own endpoints.